### ISSUE TITLE
Speeding up CustomIntegrator: When does potential energy get computed?

### ISSUE THREAD

**Issue by jchodera** [pandegroup/openmm#917]
I'm looking into potential ways to speed up `CustomIntegrator` implementations of Metropolized integrators.

If a `CustomIntegrator` calls for the potential energy at the beginning and end of a timestep, does the potential energy get computed *twice* per timestep, or just once?  For example,
```python
        # Store initial positions and velocities
        self.addComputePerDof("xold", "x")                                                                                                                                                     
        self.addComputePerDof("vold", "v")                                                                                                                                                     
        # Compute initial total energy        
        self.addComputeGlobal("Eold", "ke + energy")                                                                                                                                           
        # Symplectic velocity Verlet step
        self.addComputePerDof("v", "v + 0.5*dt*f/m")                                                                                                                                           
        self.addComputePerDof("x", "x + v*dt")                                                                                                                                                 
        self.addComputePerDof("x1", "x")                                                                                                                                                       
        self.addConstrainPositions();                                                                                                                                                          
        self.addComputePerDof("v", "v + 0.5*dt*f/m + (x-x1)/dt")                                                                                                                               
        self.addConstrainVelocities();     
        # Compute new total energy                                                                                                                                                    
        self.addComputeSum("ke", "0.5*m*v*v")                                                                                                                                                  
        self.addComputeGlobal("Enew", "ke + energy")    
        # Metropolis-Hastings acceptance/rejection step.
        self.addComputeGlobal("accept", "step(exp(-(Enew-Eold)/kT) - uniform)")                                                                                                                
        self.addComputePerDof("x", "x*accept + xold*(1-accept)")                                                                                                                               
        self.addComputePerDof("v", "v*accept - vold*(1-accept)")  
```
If twice, I wonder if there would be value in adding a new kind of `CustomIntegrator` step that is only executed once at the beginning of a series of steps, or perhaps only after the positions or velocities are updated or the parameters in the context are adjusted.  That way, we could compute and store the potential energy only once per step, using the last cached value instead of computing the initial energy.

Other useful global or per-dof variables (such as effective diffusion constants) could be computed this way too.  Perhaps via `addInitializePerDof()` and `addInitializeGlobal()` functions in analog to the existing `addCompute...()` functions.

**Comment by kyleabeauchamp** [comment]
IMHO here is my proposed plan for HMC / CustomIntegrator work:

1.  Figure out all the gory details necessary for quantitative HMC (#906, #907)
2.  Figure out what will be needed for combining fancy integrators with NPT (#895)
3.  Do some quantitative profiling to figure out the rate limiting steps
4.  Have Peter look at our code and the results of (3).  

I only mention this because it might make sense to punt on speed until after we've solved the correctness and design questions.

**Comment by jchodera** [comment]
Sounds good.  We'll eventually need to circle back to this, though.

My current understanding is that correctness may also require
* deterministic force computation (related to #216, though we don't have a separate issue open specifically covering just deterministic force and energy evaluations)
* the capability to implement bitwise reversible integrators (#916), which requires an operation to round to the nearest integer or (even better) number of signifiant bits

**Comment by kyleabeauchamp** [comment]
AFAIK aren't lots of people doing quality HMC without using fully bitwise reversible integrators?  AFAIK Gelman does not mention this in his Stan package.  Nor in the NUTS paper, per Patrick's suggestion.

**Comment by jchodera** [comment]
They are probably not using single precision on GPUs.  We're living on the bleeding edge of numerical stability, rather than the comfortable double precision world where numerical errors are minimal.

**Comment by peastman** [comment]
> If a CustomIntegrator calls for the potential energy at the beginning and end of a timestep, does the potential energy get computed twice per timestep, or just once?

It *should* only do it once, though you may want to verify that with your own code.  There are issues that can complicate it, but it does make some efforts to avoid unnecessary recomputations in this case.

**Comment by jchodera** [comment]
Closing this now.  Will reopen if we still see issues.
